Fishing bobber projectile impact

AwesomeSyd239 opened the pull request:
Adds a ProjectileImpactEvent for the FishingBobberEntity, since it extends from ProjectileEntity. 
Includes appropriate hook and post from within FishingBobberEntity class.

Also includes a simple mod showing some use cases: sticking the bobber to a block and putting the punch enchantment on a fishing rod.

I think I've done this right

AwesomeSyd239 committed:
Fishing Bobber Projectile Impact Event

Adds a ProjectileImpactEvent for the FishingBobberEntity, since it extends from ProjectileEntity. Includes appropriate hook and post from within FishingBobberEntity class.

Also includes a simple mod showing some use cases (mainly to stick the bobber to blocks).

AwesomeSyd239 committed:
Fishing Bobber Projectile Impact Event

Adds a ProjectileImpactEvent for the FishingBobberEntity, since it extends from ProjectileEntity. Includes appropriate hook and post from within FishingBobberEntity class.

Also includes a simple mod showing some use cases (mainly to stick the bobber to blocks).
2nd attempt, now with code formatted right. I'm new to this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=7842) <br/>All committers have signed the CLA.

SizableShrimp commented on src/main/resources/META-INF/mods.toml:
```diff
@@ -3,7 +3,6 @@ loaderVersion="[24,]"
 issueTrackerURL="http://www.minecraftforge.net/"
 logoFile="forge_logo.png"
 license="LGPL v2.1"
-
```
You only changed this file with whitespace.

AwesomeSyd239 replied:
I previously had fishing_example in this toml. When I moved it to src/test, there appears to be a leftover whitespace. Is there a way to completely remove the mods.toml from the PR here?

SizableShrimp replied:
The changes to this file still need to be discarded.

AwesomeSyd239 replied:
I ended up editting the file in hex to make it the same as the one from the main github. If it isn't good now, it isn't going to be.

SizableShrimp commented on src/test/java/awesomesyd/fishingexample/BarbedEnchantment.java:
```diff
@@ -0,0 +1,31 @@
+package awesomesyd.fishingexample;
```
No, you cannot have your own name and package for the tests. And a single test mod should all be in one file. Also, you need to run `licenseFormat` so that the license is applied.

AwesomeSyd239 replied:
cleaned up package stuff.
ran gradlew licenseFormat, it also put the forge license on a few things that I never touched.

SizableShrimp commented on src/main/java/net/minecraftforge/event/entity/ProjectileImpactEvent.java:
```diff
@@ -125,4 +126,23 @@ public FireworkRocketEntity getFireworkRocket()
             return fireworkRocket;
         }
     }
+    /**
```
Please update these javadocs and remove the TODO. If the event is cancelled, then the raytrace hit for the fishing hook will not be processed and it will continue moving, similar to the other projectile impact events.

AwesomeSyd239 replied:
Is it preferable to remove the doc line, as in most other ImpactEvents, or have it be the same as firework?

SizableShrimp replied:
Rewording the existing one for fireworks should be fine.

SizableShrimp commented on .gitignore:
```diff
@@ -44,3 +44,4 @@ changelog.txt
 # Srg2Source causes import bugs on these patches. If you intentionnally change these, remove the ignore.
 /patches/minecraft/net/minecraft/client/renderer/ViewFrustum.java.patch
 /patches/minecraft/net/minecraft/data/BlockModelDefinition.java.patch
+*.xml
```
Don't change the `.gitignore` file for no reason. If you need to ignore something in git, use `.git/info/exclude` so that is it not committed publicly.

AwesomeSyd239 replied:
I was trying to not include eclipseformatter.xml in my commit. Am now going with 'discard changes' on that file.

SizableShrimp reviewed (changes requested):
There a few files you changed for no good reason, and your test mod is overly complex and not in the right package. Test mods are just used to simply demonstrate stuff and/or make sure it works. They should not need a bunch of classes and enchantments.

AwesomeSyd239 committed:
Cleanup

Stripped example mod down to a single demonstration of use. Moved it to src>test-java>(default package).

Hopefully fixed whitespace issues.

Ran gradlew licenseFormat, which for some reason affected EntityTeleportEvent and a few other random files?

SizableShrimp commented on src/test/java/FishingExample.java:
```diff
@@ -0,0 +1,38 @@
+
```
Do not put the test mod in the default package. Please follow the existing examples on how to structure the package.

SizableShrimp commented on src/test/java/FishingExample.java:
```diff
@@ -0,0 +1,38 @@
+    {
+        MinecraftForge.EVENT_BUS.register(this);
+        MinecraftForge.EVENT_BUS.addListener(FishingExample::handleImpact);
+    }
```
Please add an ENABLED boolean which if true runs everything in the constructor. See other test mods for an example.

SizableShrimp reviewed (changes requested):
Looks pretty good now. Just a few comments from reading the code.

AwesomeSyd239 committed:
Copy/paste fresh toml to fix

AwesomeSyd239 committed:
cleanup

standardized naming convention, added ENABLE, and moved test file to more appropriate location.

AwesomeSyd239 committed:
Merge branch 'fishing_bobber_projectile_impact' of https://github.com/AwesomeSyd239/MinecraftForge into fishing_bobber_projectile_impact

AwesomeSyd239 committed:
Update mods.toml

AwesomeSyd239 committed:
press backspace at end of file

AwesomeSyd239 committed:
two backspace toml

AwesomeSyd239 committed:
delete whole last line of whitespace

ScLohaifg commented:
merge?

diesieben07 commented on src/main/java/net/minecraftforge/event/entity/ProjectileImpactEvent.java:
```diff
@@ -125,4 +126,24 @@ public FireworkRocketEntity getFireworkRocket()
+     * Event is cancellable, causes bobber to ignore the current hit and continue on its journey.
+     */
+    @Cancelable
+    public static class FishingBobber extends ProjectileImpactEvent//TODO figure out what happens on cancel
```
Why is this `TODO` still here?

AwesomeSyd239 replied:
because I was apparently blind. fixed now.

diesieben07 commented on src/main/java/net/minecraftforge/event/entity/living/EntityTeleportEvent.java:
```diff
@@ -1,3 +1,22 @@
+/*
```
Don't change files unrelated to your PR.

diesieben07 commented on src/main/resources/META-INF/mods.toml:
```diff
@@ -15,4 +15,3 @@ license="LGPL v2.1"
     description='''
     Forge, a broad compatibility API.
     '''
-    
```
Don't change files unrelated to your PR.

diesieben07 commented on src/test/java/net/minecraftforge/debug/block/CustomSignsTest.java:
```diff
@@ -1,3 +1,22 @@
+/*
```
Don't change files unrelated to your PR.

diesieben07 commented on src/test/java/net/minecraftforge/debug/client/rendering/LinearTextTextureFilteringTest.java:
```diff
@@ -1,6 +1,6 @@
 /*
  * Minecraft Forge
- * Copyright (c) 2016-2020.
+ * Copyright (c) 2016-2021.
```
Don't change files unrelated to your PR.

diesieben07 commented on src/test/java/net/minecraftforge/debug/entity/EntityTeleportEventTest.java:
```diff
@@ -1,3 +1,22 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
```
Don't change files unrelated to your PR.

AwesomeSyd239 committed:
Found the undo button

used git checkout to undo most of my accidental changes. mods.toml doesn't have a previous commit, but I editted it in hex to make it the same as the one on MinecraftForge main branch. If this doesn't work, I think I will have to restart the PR.
Removed a TODO line from my intended PR file.

diesieben07 commented on patches/minecraft/net/minecraft/entity/projectile/FishingBobberEntity.java.patch:
```diff
@@ -1,5 +1,14 @@
+    private void func_190624_r() {
+       RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_);
+-      this.func_70227_a(raytraceresult);
++      if (!net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) this.func_70227_a(raytraceresult);
```
All other projectiles only fire this event if it isn't a `MISS`. This fires the event even if it is a miss.

AwesomeSyd239 replied:
I think I left that out originally because most other projectiles fire the event from inside their onHit, and I don't think I realized the non-miss check was a forge insert. I was worried about accidentally messing with/cancelling things too early or accidentally.
Fixed now though.

diesieben07 replied:
Well, you are right, `func_70227_a` should still be called on `MISS`, which your patch now changes. You need an `||`, not `&&`. You want to call `func_70227_a` if it's a `MISS` _or_ the forge event was not cancelled:
`if (type == MISS || !onProjectileImpact) func_70227_a(...)`. Due to short-circuiting of `||` this will only fire the event for non-miss events still.

AwesomeSyd239 replied:
Oh, that's a good use of the short-circuit, thanks! Patched now.

AwesomeSyd239 committed:
Added non-miss condition

AwesomeSyd239 committed:
No longer breaks early on MISS

LexManos committed:
Merge branch '1.16.x' into fishing_bobber_projectile_impact

LexManos commented on patches/minecraft/net/minecraft/entity/projectile/FishingBobberEntity.java.patch:
```diff
@@ -1,5 +1,14 @@
+    private void func_190624_r() {
+       RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_);
+-      this.func_70227_a(raytraceresult);
++      if (raytraceresult.func_216346_c() == RayTraceResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) this.func_70227_a(raytraceresult);
```
This should be above the function call, so its more obvious what the patch is doing. And the line is less wide.
